File: build-tools/codeassist-builder-plugin/src/main/java/com/tyron/builder/gradle/internal/ide/BaseConfigImpl.java
package com.tyron.builder.gradle.internal.ide;

import com.android.annotations.NonNull;
import com.android.annotations.Nullable;
import com.android.annotations.concurrency.Immutable;
import com.tyron.builder.model.BaseConfig;
import com.tyron.builder.model.ClassField;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;
import java.io.File;
import java.io.Serializable;
import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.Objects;

/**
 * An implementation of BaseConfig specifically for sending as part of the Android model
 * through the Gradle tooling API.
 */
@Immutable
abstract class BaseConfigImpl implements BaseConfig, Serializable {

    @Nullable
    private final String mApplicationIdSuffix;
    @Nullable
    private final String mVersionNameSuffix;
    @NonNull
    private final Map<String, Object> mManifestPlaceholders;
    @NonNull
    private final Map<String, ClassField> mBuildConfigFields;
    @NonNull
    private final Map<String, ClassField> mResValues;
    @Nullable
    private final Boolean mMultiDexEnabled;
    @Nullable
    private final File mMultiDexKeepFile;
    @Nullable
    private final File mMultiDexKeepProguard;
    @NonNull private final List<File> mProguardFiles;
    @NonNull private final List<File> mConsumerProguardFiles;
    @NonNull private final List<File> mTestProguardFiles;

    protected BaseConfigImpl(@NonNull BaseConfig baseConfig) {
        mApplicationIdSuffix = baseConfig.getApplicationIdSuffix();
        mVersionNameSuffix = baseConfig.getVersionNameSuffix();
        mManifestPlaceholders = ImmutableMap.copyOf(baseConfig.getManifestPlaceholders());
        mBuildConfigFields = ImmutableMap.copyOf(baseConfig.getBuildConfigFields());
        mResValues = ImmutableMap.copyOf(baseConfig.getResValues());
        mMultiDexEnabled = baseConfig.getMultiDexEnabled();
        mMultiDexKeepFile = baseConfig.getMultiDexKeepFile();
        mMultiDexKeepProguard = baseConfig.getMultiDexKeepProguard();
        mProguardFiles = ImmutableList.copyOf(baseConfig.getProguardFiles());
        mConsumerProguardFiles = ImmutableList.copyOf(baseConfig.getConsumerProguardFiles());
        mTestProguardFiles = ImmutableList.copyOf(baseConfig.getTestProguardFiles());
    }

    @Nullable
    @Override
    public String getApplicationIdSuffix() {
        return mApplicationIdSuffix;
    }

    @Nullable
    @Override
    public String getVersionNameSuffix() {
        return mVersionNameSuffix;
    }

    @NonNull
    @Override
    public Map<String, ClassField> getBuildConfigFields() {
        return mBuildConfigFields;
    }

    @NonNull
    @Override
    public Map<String, ClassField> getResValues() {
        return mResValues;
    }

    @NonNull
    @Override
    public List<File> getProguardFiles() {
        return mProguardFiles;
    }

    @NonNull
    @Override
    public List<File> getConsumerProguardFiles() {
        return mConsumerProguardFiles;
    }

    @NonNull
    @Override
    public Collection<File> getTestProguardFiles() {
        return mTestProguardFiles;
    }

    @Override
    @NonNull
    public Map<String, Object> getManifestPlaceholders() {
        return mManifestPlaceholders;
    }

    @Override
    @Nullable
    public Boolean getMultiDexEnabled() {
        return mMultiDexEnabled;
    }

    @Nullable
    @Override
    public File getMultiDexKeepFile() {
        return mMultiDexKeepFile;
    }

    @Nullable
    @Override
    public File getMultiDexKeepProguard() {
        return mMultiDexKeepProguard;
    }

    @Override
    public String toString() {
        return "BaseConfigImpl{"
                + "applicationIdSuffix='"
                + mApplicationIdSuffix
                + '\''
                + ", versionNameSuffix='"
                + mVersionNameSuffix
                + '\''
                + ", mManifestPlaceholders="
                + mManifestPlaceholders
                + ", mBuildConfigFields="
                + mBuildConfigFields
                + ", mResValues="
                + mResValues
                + ", mMultiDexEnabled="
                + mMultiDexEnabled
                + ", mProguardFiles="
                + mProguardFiles
                + ", mConsumerProguardFiles="
                + mConsumerProguardFiles
                + ", mTestProguardFiles="
                + mTestProguardFiles
                + '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        BaseConfigImpl that = (BaseConfigImpl) o;
        return Objects.equals(mApplicationIdSuffix, that.mApplicationIdSuffix)
                && Objects.equals(mVersionNameSuffix, that.mVersionNameSuffix)
                && Objects.equals(mManifestPlaceholders, that.mManifestPlaceholders)
                && Objects.equals(mBuildConfigFields, that.mBuildConfigFields)
                && Objects.equals(mResValues, that.mResValues)
                && Objects.equals(mMultiDexEnabled, that.mMultiDexEnabled)
                && Objects.equals(mMultiDexKeepFile, that.mMultiDexKeepFile)
                && Objects.equals(mMultiDexKeepProguard, that.mMultiDexKeepProguard)
                && Objects.equals(mProguardFiles, that.mProguardFiles)
                && Objects.equals(mConsumerProguardFiles, that.mConsumerProguardFiles)
                && Objects.equals(mTestProguardFiles, that.mTestProguardFiles);
    }

    @Override
    public int hashCode() {
        return Objects.hash(
                mApplicationIdSuffix,
                mVersionNameSuffix,
                mManifestPlaceholders,
                mBuildConfigFields,
                mResValues,
                mMultiDexEnabled,
                mMultiDexKeepFile,
                mMultiDexKeepProguard,
                mProguardFiles,
                mConsumerProguardFiles,
                mTestProguardFiles);
    }
}

File: build-tools/codeassist-builder-plugin/src/main/java/com/tyron/builder/internal/BaseConfigImpl.kt
package com.tyron.builder.internal

import com.google.common.collect.Lists
import com.google.common.collect.Maps
import com.tyron.builder.model.BaseConfig
import com.tyron.builder.model.ClassField
import java.io.File
import java.io.Serializable

/**
 * An object that contain a BuildConfig configuration
 */
abstract class BaseConfigImpl : Serializable,
    BaseConfig {

    override var applicationIdSuffix: String? = null

    open fun applicationIdSuffix(applicationIdSuffix: String?) {
        this.applicationIdSuffix = applicationIdSuffix
    }

    override var versionNameSuffix: String? = null

    open fun versionNameSuffix(versionNameSuffix: String?) {
        this.versionNameSuffix = versionNameSuffix
    }

    private val mBuildConfigFields: MutableMap<String, ClassField> =
        Maps.newTreeMap()
    private val mResValues: MutableMap<String, ClassField> =
        Maps.newTreeMap()
    private val mProguardFiles: MutableList<File> =
        Lists.newArrayList()
    private val mConsumerProguardFiles: MutableList<File> =
        Lists.newArrayList()
    private val mTestProguardFiles: MutableList<File> =
        Lists.newArrayList()
    private val mManifestPlaceholders: MutableMap<String, Any> =
        Maps.newHashMap()

    override var multiDexEnabled: Boolean? = null

    override var multiDexKeepProguard: File? = null

    override var multiDexKeepFile: File? = null

    /**
     * @see .getApplicationIdSuffix
     */
    fun setApplicationIdSuffix(applicationIdSuffix: String?): BaseConfigImpl {
        this.applicationIdSuffix = applicationIdSuffix
        return this
    }

    /**
     * @see .getVersionNameSuffix
     */
    fun setVersionNameSuffix(versionNameSuffix: String?): BaseConfigImpl {
        this.versionNameSuffix = versionNameSuffix
        return this
    }

    /**
     * Adds a BuildConfig field.
     */
    fun addBuildConfigField(field: ClassField) {
        mBuildConfigFields[field.name] = field
    }

    /**
     * Adds a generated resource value.
     */
    fun addResValue(resValueKey: String, field: ClassField) {
        mResValues[resValueKey] = field
    }

    /**
     * Adds a generated resource value.
     */
    fun addResValues(values: Map<String, ClassField>) {
        mResValues.putAll(values)
    }

    override var buildConfigFields: Map<String, ClassField>
        get() = mBuildConfigFields
        set(fields) {
            mBuildConfigFields.clear()
            mBuildConfigFields.putAll(fields)
        }

    /**
     * Adds BuildConfig fields.
     */
    fun addBuildConfigFields(fields: Map<String, ClassField>) {
        mBuildConfigFields.putAll(fields)
    }

    /**
     * Returns the generated resource values.
     */
    override var resValues: MutableMap<String, ClassField>
        get() = mResValues
        set(fields) {
            mResValues.clear()
            mResValues.putAll(fields)
        }

    /** {@inheritDoc}  */
    override val proguardFiles: MutableList<File>
        get() = mProguardFiles

    override val consumerProguardFiles: MutableList<File>
        get() = mConsumerProguardFiles

    override val testProguardFiles: MutableList<File>
        get() = mTestProguardFiles

    override val manifestPlaceholders: MutableMap<String, Any>
        get() = mManifestPlaceholders

    fun addManifestPlaceholders(manifestPlaceholders: Map<String, Any>) {
        mManifestPlaceholders.putAll(manifestPlaceholders)
    }

    fun setManifestPlaceholders(manifestPlaceholders: Map<String, Any>): Void? {
        mManifestPlaceholders.clear()
        mManifestPlaceholders.putAll(manifestPlaceholders)
        return null
    }

    // Here to stop the gradle decorator breaking on the duplicate set methods.
    open fun manifestPlaceholders(manifestPlaceholders: Map<String, Any>) {
        setManifestPlaceholders(manifestPlaceholders)
    }

    open fun _initWith(that: BaseConfig) {
        buildConfigFields = that.buildConfigFields
        resValues = that.resValues as MutableMap<String, ClassField>
        applicationIdSuffix = that.applicationIdSuffix
        versionNameSuffix = that.versionNameSuffix
        mProguardFiles.clear()
        mProguardFiles.addAll(that.proguardFiles)
        mConsumerProguardFiles.clear()
        mConsumerProguardFiles.addAll(that.consumerProguardFiles)
        mTestProguardFiles.clear()
        mTestProguardFiles.addAll(that.testProguardFiles)
        mManifestPlaceholders.clear()
        mManifestPlaceholders.putAll(that.manifestPlaceholders)
        multiDexEnabled = that.multiDexEnabled
        multiDexKeepFile = that.multiDexKeepFile
        multiDexKeepProguard = that.multiDexKeepProguard
    }

    override fun toString(): String {
        return "BaseConfigImpl{" +
                "applicationIdSuffix=" + applicationIdSuffix +
                ", versionNameSuffix=" + versionNameSuffix +
                ", mBuildConfigFields=" + mBuildConfigFields +
                ", mResValues=" + mResValues +
                ", mProguardFiles=" + mProguardFiles +
                ", mConsumerProguardFiles=" + mConsumerProguardFiles +
                ", mManifestPlaceholders=" + mManifestPlaceholders +
                ", mMultiDexEnabled=" + multiDexEnabled +
                ", mMultiDexKeepFile=" + multiDexKeepFile +
                ", mMultiDexKeepProguard=" + multiDexKeepProguard +
                '}'
    }

    companion object {
        private const val serialVersionUID = 1L
    }
}